Reprogrammed the normalizer and langs functionality

m3hrdadfi opened the pull request:
Hi @anton-l 

🤩 That's really a good idea to create a toolkit, and I'm willing to contribute for the sake of all possible languages. I reprogrammed the normalizer in order to bring the functionality of text preprocessing for every language separately. It will be easy for every native speaker to add characters to filter or map or even a normalizing and preprocessing procedure at the word or text level.

m3hrdadfi committed:
Reprogrammed the normalizer and langs functionality

m3hrdadfi committed:
Added summary to normalizer

m3hrdadfi committed:
Reformated

anton-l commented on src/wav2vec_toolkit/text_preprocessing/lang/default.py:
```diff
@@ -0,0 +1,48 @@
+CHARS_TO_IGNORE = [
```
IMO, whitelisting characters (something like `ALPHABET`) would work better than blacklisting with `CHARS_TO_IGNORE`. That way you won't have to sift through all types of unicode punctuation to compile this list.

anton-l replied:
The only downside is that this would be a vary large list for non alphabet-based languages. But it can be implemented as a json path, perhaps?

m3hrdadfi replied:
You're right. Whitelist is more sensible and easy to find for every language. I have something in my mind, suppose that we created a whitelist regex dictionary for every language and try to tokenize the text to select the correct possible characters, more like the following snippet.

```python
import re


whitelist_regexes = {
    "en": r"[A-Za-z]+",
    "tr": r"[a-zA-ZçÇğĞüÜöÖşŞıİ]+",
    "fa": r"[ءآئابتثجحخدذرزسشصضطظعغفقلمنهوپچژکگیە\u200c]+"
}

def tokenizer(text, lang):
    pattern = whitelist_regexes[lang]
    tokenized = re.findall(pattern, text)
    return " ".join(tokenized)
```

```python
text = "Twitter, 2021'deki en iyi sosyal platformdur."
print(f"BEFORE: {text}")
print(f"AFTER: {tokenizer(text, 'tr')}")
```

Output:
```text
BEFORE: Twitter, 2021'deki en iyi sosyal platformdur.
 AFTER: Twitter deki en iyi sosyal platformdur
```

```python
text = "Twitter, 2021'deki en iyi sosyal platformdur. این یک زبان دیگست"
print(f"BEFORE: {text}")
print(f" AFTER: {tokenizer(text, 'tr')}")
```

Output:
```text
BEFORE: Twitter, 2021'deki en iyi sosyal platformdur. این یک زبان دیگست
 AFTER: Twitter deki en iyi sosyal platformdur
```

anton-l replied:
Yeah, this would work! You can also use `re.IGNORECASE` so that you don't have to include uppercase characters in the regex patterns.

m3hrdadfi replied:
Ok then, I'll change the required sections based on this scenario in the next commit.

anton-l commented on src/wav2vec_toolkit/text_preprocessing/normalizer.py:
```diff
@@ -0,0 +1,159 @@
+    return sentence
+
+
+def word_level_normalizer(sentence: str, fn_callback=None) -> str:
```
Word-level actions seem redundant if we already have text-level normalization. And this is especially relevant for laguages that can't be tokenized with a simple `sentence.split(" ")`. I suggest we leave only text-level normalization and let everyone tokenize the text if needed for some specific steps.

m3hrdadfi replied:
Maybe it's true overall, but in some languages, the speaker mentioned something different from what they entered. For example, the speaker may say a number in a way other than writing.

SAID: one hundred and twenty-five
ENTERED: 125

In those cases, we need to handle possible solutions at the word level, like the following snippet.

```python
def int2str(text):
    raise NotImplemented


def word_level_action(word):

    # 125 -> one hundred and twenty-five
    word = int2str(word)

    return word
```

anton-l replied:
Good point! Bust still, this use case would need some universal tokenization (that also handles Chinese, Japanese, etc) before applying the word-level function. But if we leave tokenization to the user, then `text_level_action` alone would be enough to cover all normalization, what do you think?

m3hrdadfi replied:
Correct, in the case of Chinese, Japanese, or other similar, might become so difficult and redundant. I agree to stick to the `text_level_action` 😄

anton-l commented on src/wav2vec_toolkit/text_preprocessing/normalizer.py:
```diff
@@ -0,0 +1,159 @@
+    if remove_extra_space:
+        text = re.sub(" +", " ", text)
+
+    if remove_last_space:
```
I remember a discussion on slack about whether we need the space, what was the outcome? Can't find the thread, unfortunately.

m3hrdadfi replied:
If we use the whitelist tokenizer, we don't need to handle extra spaces anymore.

borisdayma replied:
From Patrick, we needed the extra space for training but not for testing (see [conversation on slack](https://huggingface.slack.com/archives/C01QZ90Q83Z/p1616426911010100)).
Though I'm not exactly sure why.

anton-l commented on src/wav2vec_toolkit/text_preprocessing/normalizer.py:
```diff
@@ -0,0 +1,159 @@
+
+    # Remove extra spaces
+    if remove_extra_space:
+        text = re.sub(" +", " ", text)
```
`re.sub("\s+", " ", text)` will handle more types of whitespace

anton-l reviewed (changes requested):
Thank you @m3hrdadfi ! This is very close to what I've imagined would be the most plug-and-play solution :+1:

m3hrdadfi committed:
Reprogrammed the normalizer #5

m3hrdadfi committed:
Reprogrammed the normalizer #5

m3hrdadfi committed:
Reprogrammed the normalizer #5

m3hrdadfi commented:
I checked the scenario one more time, and some steps need to be taken into account. 

- We need an easy procedure for adding new languages without affecting each other.
- We need a solution to handle requirements related to each language-normalization standalone; for instance, in Farsi, we use Hazm or Parsivar.
- We need to add some parameters grabbing to present specific language params or fine-tuning arguments. 
- Also, some documentation regarding each language to convey a better understanding of that specific language.

It's not the ultimate solution but might be a starting point. What do you think, @anton-l?

How to use

```python
from wav2vec_toolkit.text_preprocessing.normalizers import normalizers
from wav2vec_toolkit.text_preprocessing.parameters import parameters

Normalizer = normalizers("tr")
Parameter = parameters("tr")

text = "Twitter, 2021'deki en iyi sosyal platformdur. این مربوط به یه زبان دیگست" 
batch = {
    "text": text
}
normalizer = Normalizer(
    do_lowercase=False,
    text_key_name="text",
    dictionary={"2021": "two thousand twenty-one"}
)
print(f"Normalized: {normalizer(batch)}")
print(f"  Language: {Parameter.language_name}")
print(f"      Code: {Parameter.language_code}")
```

Output:
```text
Normalized: {'text': 'Twitter two thousand twenty-one deki en iyi sosyal platformdur'}
  Language: Turkish
      Code: tr
```

m3hrdadfi committed:
Reprogrammed the normalizer #5

anton-l commented on src/wav2vec_toolkit/utils.py:
```diff
@@ -0,0 +1,83 @@
+
+
+def load_module_from_lang(lang: str):
+    lang_mod_path = f"{BASE_PATH}/{LANG_PATH}/{lang}"
```
Have you checked that this works when the package is installed via `pip`?

m3hrdadfi replied:
Not actually, but I've checked now and found the bug and fixed it.

m3hrdadfi replied:
I've just added the parameters for storing the hyperparams, but since @borisdayma uses the hyperparam_search.py, there's no need to add this part.

m3hrdadfi replied:
@anton-l, I'll update some minor bugs in the next commit, improve some info, and add the instruction like what Huggingface-datasets did.

anton-l reviewed (changes requested):
Great work @m3hrdadfi ! 
Regarding the `parameters` module, I think it would be sufficient to just keep the hyperparams inside the finetuning bash scripts for each language. @borisdayma offered to help with setting up the entrypoints for the scripts with wandb tracking https://github.com/anton-l/wav2vec-toolkit/issues/4#issuecomment-814593222 
So having just the `normalizer`s is enough for now.

As for the other points, your solution for them so far looks very flexible and clean, I have no issues with it! :+1: 

Can you please add a little snippet to the main README, detailing what steps should should the contributors take to integrate a new language into the library? Then we can merge this PR and try adding some new languages :rocket:

borisdayma commented:
Not sure if related or not but at this time I'm experimenting with keeping the punctuation to see if it could also be predicted.

Regarding wandb tracking, I imagined it would be done in `wav2vec-toolkit/src/wav2vec_toolkit/finetune.py`.
Then `https://github.com/anton-l/wav2vec-toolkit/blob/main/scripts/hyperparam_search.py` would just call that script and configuration would be done through a YAML file (actually I'm not sure we would need a script at all).

anton-l commented:
@borisdayma IIRC, punctuation is usually handled via LM post-processing (to make the STT's model targets less noisy and more sound-dependent), but it's an interesting problem to tackle nevertheless.

As for the tracking, I thought the same, yaml/json config parsing would be perfect (and already mostly implemented in huggingface scripts).

m3hrdadfi committed:
Adding potential to extend new languages

m3hrdadfi committed:
Adding potential to extend new languages

m3hrdadfi commented:
Hi @anton-l,

I've changed some critical parts:

- Isolated the language's directory from the main src code.
- Added guide as `ADD_NEW_LANGUAGE.md` on how to add a new language as well as updated the main `README.md`.
- Removed the stored parameters from the language directory.
- Added a template regarding adding a new language.
- Updated the `setup.py` and added `MANIFEST.in` in order to handle the external languages directory.

anton-l commented:
@m3hrdadfi Awesome! The package structure really comes together, let's merge everything now :)
